openssl-sys/build.rs

```diff
@@ -1,86 +1,323 @@
+        if system.exists() {
+            panic!("
+
+Failed to find an approprate OpenSSL installation. It looks like you're
+compiling on macOS, where the system contains a version of OpenSSL 0.9.8. This
+crate no longer supports OpenSSL 0.9.8.
+
+As a consumer of this crate, you can fix this error by using Homebrew to
+install the `openssl` package, or as a maintainer you can use the openssl-sys
+0.7 crate for support with OpenSSL 0.9.8.
+
+Unfortunately though the compile cannot continue, so aborting.
+
+");
         }
     }
 
-    let libs_env = env::var("OPENSSL_LIBS").ok();
-    let libs = match libs_env {
-        Some(ref v) => v.split(":").collect(),
-        None => if target.contains("windows") {
-            if get_mingw_in_path().is_some() && lib_dir.is_none() && include_dir.is_none() {
-                vec!["ssleay32", "eay32"]
-            } else {
-                vec!["ssl32", "eay32"]
-            }
-        } else {
-            vec!["ssl", "crypto"]
+    try_pkg_config();
+
+    // TODO: try to be more helpful in this message, e.g.:
+    //
+    // * if we're on linux, recommend apt-get something or other
+    // * if we're on Windows, link to a README explaining more
+
+    panic!("
+
+Could not find directory of OpenSSL installation, and this `-sys` crate cannot
+proceed without this knowledge. To help this crate find this directory
+you can set the `OPENSSL_DIR` environment variable for the compilation process.
+
+If you're in a situation where you think the directory *should* be found
+automatically, please open a bug at https://github.com/sfackler/rust-openssl
+and include information about your system as well as this message.
+
+    $HOST = {}
+    $TARGET = {}
+    openssl-sys = {}
+
+",
+    host, target, env!("CARGO_PKG_VERSION"));
+}
+
+/// Attempt to find OpenSSL through pkg-config.
+///
+/// Note that if this succeeds then the function does not return as pkg-config
+/// typically tells us all the information that we need.
+fn try_pkg_config() {
+    let target = env::var("TARGET").unwrap();
+    let host = env::var("HOST").unwrap();
+
+    // If we're going to windows-gnu we can use pkg-config, but only so long as
+    // we're coming from a windows host.
+    //
+    // Otherwise if we're going to windows we probably can't use pkg-config.
+    if target.contains("windows-gnu") && host.contains("windows") {
+        env::set_var("PKG_CONFIG_ALLOW_CROSS", "1");
+    } else if target.contains("windows") {
+        return
+    }
+
+    // We're going to be looking at header files, so show us all the system
+    // cflags dirs for showing us lots of `-I`.
+    env::set_var("PKG_CONFIG_ALLOW_SYSTEM_CFLAGS", "1");
+
+    let lib = match pkg_config::find_library("openssl") {
+        Ok(lib) => lib,
+        Err(_) => return,
+    };
+
+    if lib.include_paths.len() == 0 {
+        panic!("
+
+Used pkg-config to discover the OpenSSL installation, but pkg-config did not
```

Comment:
In theory `PKG_CONFIG_ALLOW_SYSTEM_CFLAGS` set above will force `pkg-config` to print all those out, but if that doesn't work then yeah this'll be a weird failure.